### ISSUE TITLE
Fix relative paths in playlists

### ISSUE THREAD

**Issue by dragonmaus** [vanilla-music/vanilla#981]
The relative paths currently stored in playlist files are very wrong.

Given the following:

- Playlist file:
  `/storage/emulated/0/Music/Favorites.m3u`
- Music structure:
  `/storage/emulated/0/Music/<artist>/<album>/<track>`

I would expect the playlist file to contain:

```m3u
<artist>/<album>/<track>
```

Instead it contains:

```m3u
../../storage/emulated/0/Music/<artist>/<album>/<track>
```

which is A) not useful (since the whole point of relative paths is to *remove* references to the common parent directories [in this case `/storage/emulated/0/Music`]), and B) completely incorrect (it would resolve to `/storage/emulated/storage/emulated/0/Music/<artist>/<album>/<track>`, which does not exist). 

I have confirmed this to be the case no matter where the playlist is stored (e.g. `/storage/emulated/0/Sync/Playlists/Favorites.m3u` returns `../../../storage/emulated/0/Music/<artist>/<album>/<track>` when I would expect `../../Music/<artist>/<album>/<track>`).

**Comment by adrian-bl** [comment]
Playlist synchronization defaults to /sdcard/Playlists which would make your paths the correct relative path.

What path do you have configured under Settings >> Playlists?

**Comment by dragonmaus** [comment]
I could have been more explicit, I guess.

`/sdcard` is a symlink to `/storage/emulated/0` on my device (which reveals *another* issue with the relative path function), and I had my playlist folder set to `/storage/emulated/0/Music`.

However, that should not matter, and would not if the relative path function provided a *useful* response. As I indicated in my original comment, the entire purpose of including this feature is to remove things like `/storage/emulated/0` or `/sdcard`, so that the playlist file works on multiple machines.

TL;DR: In order to be correct and useful, a relative path function must:
- Calculate the path relative to the actual playlist file involved,
- Canonicalize all input paths, and
- Provide the shortest possible result.

Descending to `/` using `..` and then appending the absolute path to the target is *not* a valid technique.

**Comment by dragonmaus** [comment]
Here is a simple shell function showing a correct `relative_path` function:

```sh
relative_path() {
	local from="$(realpath "$1")"
	local to="$(realpath "$2")"
	local ascend=
	[[ -d "$from" ]] || from="$(dirname "$from")"
	while [[ "$to" != "$from"/* ]]
	do
		ascend="../$ascend"
		from="$(dirname "$from")"
	done
	echo "$ascend${to#"$from/"}"
}
```

This returns the following:
- `relative_path /sdcard/Playlist/Favorite.m3u /sdcard/Music/Path/To/Song.flac`
  => `../Music/Path/To/Song.flac`
- `relative_path /sdcard/Playlist /sdcard/Music/Path/To/Song.flac`
  => `../Music/Path/To/Song.flac`
- `relative_path /storage/emulated/0/Music/Favorite.m3u /sdcard/Music/Path/To/Song.flac`
  => `Path/To/Song.flac`
- `relative_path /sdcard/Sync/Playlists /sdcard/Music/Path/To/Song.flac`
  => `../../Music/Path/To/Song.flac`

**Comment by adrian-bl** [comment]
> (which reveals another issue with the relative path function) ...

/sdcard is probably a bad default path for the playlists, but i do not think that we should resolve symlinks: After an OS upgrade, /sdcard might become a bind mount or something else: The code in vanilla can not assume that the path stays a symlink forever.

Commit aaa5f8e3ef95ba63aabbd6a69f98f2d534584209 changes the default, so it would most likely default to `/storage/emulated/0/....`

**Comment by dragonmaus** [comment]
> i do not think that we should resolve symlinks... The code in vanilla can not assume that the path stays a symlink forever.

Good point; I had not considered that angle.

That said, the default playlists path was tangential to the issue I opened; please reconsider closing this topic, as fixing the default in no way fixes the problem.

**Comment by dragonmaus** [comment]
Please re-open this issue. Commit aaa5f8e has nothing to do with the original problem I reported.

**Comment by adrian-bl** [comment]
I can not reproduce this bug: please attach a screenshot of the playlist config path and mediablibrary settings.

**Comment by dragonmaus** [comment]
(I completely wiped the app and started fresh, so you should be able to reproduce this on your end.)

### Screenshots
- Playlist settings
  ![Playlist settings](https://i.imgur.com/PWoxjCL.png)
- Playlist config path
  ![Playlist config path](https://i.imgur.com/GBYLKtq.png)
- Media library settings
  ![Media library settings](https://i.imgur.com/ld6ZFqp.png)

### Expected results
`Favorites.m3u` contains
```
../Music/Jeremy Soule/1997 - Total Annihilation/02 - Brutal Battle.flac
```

### Actual results
`Favorites.m3u` contains
```
../../storage/emulated/0/Music/Jeremy Soule/1997 - Total Annihilation/02 - Brutal Battle.flac
```

**Comment by adrian-bl** [comment]
So the issue is what i described before:

Your sync path is set to /sdcard/Playlists which is the reason you see the unexpected (but correct) relative paths.

Change the playlist sync path to /storage/emulated/0/Playlists and things should be fine.

**Comment by dragonmaus** [comment]
I feel somewhat foolish. Now that I have actually taken the time to consider what is going on, I see how the function is working, *and* how your patch above fixes it.

I apologise for my aggressive tone, and thank you for your assistance.
